ezgz.hpp

```diff
@@ -517,7 +517,7 @@ class EncodedTable {
 			}
 		}
 
-		for (int& it : codesIndex) it = UNUSED;
+		std::memset(codesIndex.data(), UNUSED, codesIndex.size());
```

Comment:
I don't think result of this is the same. I see that UNUSED is an int(-2), memset is supposed to set each byte to static_cast<unsigned char>(UNUSED)

Reply:
Wow, I completely forgot that the elements in that array are of type int. It didn't break anything because those values were there only for debugging. I will just remove it then.

Reply:
Anyway, as someone suggested std::fill somewhere, there is not really a difference here, if you look at the assembly, its the same as custom loop with gcc/clang on amd64 with O2:
https://godbolt.org/z/avWEKjohz

Reply:
Memset produces something different, but I don't understand assembly enough to tell what it does: https://godbolt.org/z/rGYoor5PW

Anyway, this line doesn't have to be there, I only needed to distinguish zero index from unset index when debugging. I will remove it.

Reply:
If you are interested, GCC output explanation:
`mov rdi, rbx` register is set to pointer to the buffer
`mov ecx, 128` this initializes "loop" counter for the `rep` instruction to 128
`movabs  rax, -72340172838076674` initializes constant to write into the buffer, this is 0xfe (-2 as unsigned char)  repeated 8 times (0xfefefefefefefefe)
`rep stosq` this is the actual loop - rax is written to where rdi points to (8 bytes), increases rdi by 8, decreases rcx by one, loop ends when rcx becomes 0;

So basicly, it fills the buffer by writing eight bytes at a time.

The manual loop and std::fill did 16 bytes at a time, by using 16B XMM register and SSE instructions.

What clang did, is that it unrolled the loop, repeating the instructions in case of manual loop and std::fill. In case of memset, it just calls the function from C library, not doing any inlining.

Reply:
Interesting.

I thought the C functions were the easiest to replace with SEE instructions, but in this assembly, it looks like the worse option. I'll have to play with it, maybe it can be used to optimise the code a bit.